### ISSUE TITLE
Textures

### ISSUE THREAD

**Issue by HellDragon-HK** [GreatEmerald/UT3Vehicles#19]
Issues with textures, I changed the format a bit, should it be reverted to check boxes?
  - [ ] Manta Textures
  - [ ] Full 64 bit compatibility
   - Front dot lights
      Red Team - Need to be slightly more red and is missing a subtle glow effect that encompasses all of the dots at once
     Blue Team - Need to be slightly darker blue and are missing a glow effect that encompasses all of the dots at once
   - Interior Controls 
      Red Team - Currently it's very bright, in UT3 it's a very dull brownish dark red that doesn't glow
      Blue Team - Currently it's bright as with Red it's a dull dark blue that doesn't glow
  - Middle bar
      Red Team - Color needs more red and is still missing a subtle glow effect
      Blue Team - Color is good, it's just missing a large glow effect now
  - Tail Lights
      Red Team - I'm inclined to say the color here is correct and is now only missing that subtle glow effect
      Blue Team - Color is good again, just missing a large glow now
  - [ ] Scorpion
  - [ ] Full 64 bit compatibility
  - Blades
     Red Team - Needs very slightly more red and 4 subtle glow effects are still missing
     Blue Team - Very slightly could use a darker blue and is missing the 4 glow effect spots
  - Headlights
     Red Team - Not quite red enough yet and don't have that intense look
     Blue Team - Could use a darker blue but it's close, glow isn't intense enough yet
  - Side Lights
     Red Team - Almost perfect, could use a little more red, if you don't want to it's fine the way it is, you even got the subtle glow here so great work
      Blue Team - In UT3 it looks like they are only slightly brighter blue than the Scorpion itself, other than that perfect.
  - Lower Tail Light
      Red Team - Could use some more red
      Blue Team - Needs to be a darker blue
  - Tail Light - Neither light should be lit up unless braking, white light is lit up while in reverse and red light is lit when actually braking
   - With 64 bit exe the lights are all there but the specular map is missing

 - [ ] Hellbender
 - [ ] Full 64 bit compatibility
  - Headlights
     Blue Team - Headlights have lost the grill pattern, missing large glow, and need to be darker blue
     Red Team - Need to be slightly darker and are missing that glow effect
  - Driver Turret - I didn't even notice driver turret had a light, good job Alex
      Blue Team - Missing it's glow still
      Red Team - The driver turret is good just needing a subtle glow effect now
  - Back Turret -  Blue Team - I see some red that isn't there in UT3 and that blue in UT3 is much more dull sky blue kind of....Red Team is almost perfect and doesn't need anything
  - Tail Lights - Red tail lights only light up on braking, and white is only when in reverse

  - [ ] Raptor
  - [ ] Full 64 bit compatibility
   - Front Lights
       Red Team - Could slightly use more red and bottom light in UT3 doesn't appear to glow or is more dull than top light
       Blue Team - Needs to be almost the same as the blue metal above it and bottom light needs slightly less glow, just like with Red the top light has more glow than the bottom one but it's less true with blue team
   - Wings
        Red Team - Only needs to be slightly more red
        Blue Team - Looks good, could use a slightly darker blue
  - Engines
        Red Team - What happened here!? There is a lot wrong here, first off the whole thing shouldn't be glowing peach red, only the 4 bar sections in the back should be, and those 4 bars are bright and glowing in the middle only (all 4 glow vertically) while outwards is either not glowing or just dull colored, and everything else is plain metal
        Blue Team - What again, same issue as with Red only for Blue it looks like it doesn't have as much variation of shades on the bars

  - [ ] Cicada
  - [ ] Full 64 bit compatilibility
     - Front Light
         Red Team - All it needs is a subtle glow effect now
         Blue Team - Needs to be slightly darker blue is all
     - Rotor Lights
         Red Team - I think the color is good and it needs a red glow added still
         Blue Team - I think the color is good again and needs a dark blue glow added
     - Gunner Turret Light
         Red Team - Needs both more red and pink as well as a glow effect
         Blue Team - Needs to be slightly darker blue and have a glow effect added

  - [ ] Hellfire SPMA
  - [ ] Full 64 bit compatibility
    - Headlights
        Red Team - This one really is peach in UT3, looks like it could use more pink and a glow effect which just like the Raptor the top has glow while the bottom one either barely does or not at all
         Blue Team - Need to be slightly darker blue and a little more glow, again top needs to glow more than bottom
    - Top Lights -  
         Blue Team - Needs to be dark blue, the wire or hose part is fairly decent as is but should probably also be darkened
     - Tail Lights - If the Hellfire does have tail lights they shouldn't be glowing at all times

  - [ ] Paladin
    - Front Lights
         Red Team - Front large light needs more red and has a glow effect above and below that is still missing, the dot lights to the side of it are almost red enough but could use a little more (this is my opinion anyway).
          Blue Team - As with Red Team the front light has a glow effect still missing, the dot lights are near perfect however
   - Turret
          Red Team - Really bright in UT2004, needs to be darker or less bright
          Blue Team - Almost perfect just needs to be more of a dark blue although is good the way it is
   - Alex, you added some tail lights to the back that shouldn't be there, I'm not going to complain if that doesn't change because I kind of like it but it is inconsistent, this is true for both teams and Paladin doesn't appear to have any tail lights in UT3
 
  - [ ] Goliath
  - [ ] Full 64 bit compatibility
  - Headlights
        Red Team - Needs slightly more red and that intense glow effect
        Blue Team - Needs slightly darker blue and a more intense glow effect
  - Turret
        Red Team - Needs more red and largest circle needs glow effect, small spot on the other side is good enough but could use slightly more red
        Blue Team - Needs darker blue on both areas, more intense glow on large circle, and has kind of lost it's grill pattern on the large circle

 - [ ] Leviathan
 - [ ] Full 64 bit compatibility
  - Headlights
        Red Team - Needs to be more red is all
        Blue Team - Need to be slightly darker blue
  - Main Cannon
        Blue Team - Needs to be slightly darker blue and has a little too much glow
  - Tail Lights - Need to be determined if they are there or not in UT3, not sure if this falls under texture or general issue since it would also need the brakelight in the code

  - [ ] Viper
  - [ ] Full 64 bit compatibility
  - Front Light
         Red Team - Needs to be a dark red and maybe bit a dark pink (also remember I'm on PS3 so my TV colors may differ and you should take a look yourself in-game first)
         Blue Team - Needs to be darker blue
  - Front Dot Lights
         Red Team - Needs to be darker red and maybe a bit of dark pink, also looks like the pulse isn't quite right, this section in UT3 has more aggressive pulse that periodically has all 3 lights with a slight intense glow, if that is too much it's still fine the way it is.
          Blue Team - Needs to be darker blue and a slight glow effect added, pulse like with Red doesn't seem quite aggressive enough yet, possibly might be slower in UT2004
   - Side Lights
           Red Team - Needs more red and pink as well as pulses like the front dot lights, a clear glow can be noticed with the single smallest dot in-between the front and side lights....maybe could be described as has a Morse code pattern for it's glow
           Blue Team - Looks good except that single small dot light which I think needs to be a dull dark blue and very seldom in UT3 it has this subtle greenish blue glow around it in it's pulse pattern
   - Controls
            Red Team - Needs to be darker red and more pink, looks like the pulse effect is almost exact here so use it as a reference to check the previous light pulse patterns
            Blue Team - Almost looks perfect, could use a slightly darker blue
   - Hose Lights
            Red Team - Needs more pink and slightly darker red
            Blue Team - Almost perfect, could use very slightly darker blue
   - Outer Wing Lights
            Red Team - Needs more dark pink and in UT3 has noticeable glow in pulse pattern that is still missing
            Blue Team - Almost perfect, could use slightly darker blue
   - Inner Wing Lights
            Red Team - Needs more dark pink and in UT3 looks like the pulse effect might be slightly faster
            Blue Team - Too much blue here, I'm seeing blue only on the top for the most part in UT3 while the rest of it is more of a almost black blue, also the top blue needs to be darker
  - Back Lights (I can't see all of these very well in UT3 due to collision)
            Red Team - Looks good as far as I can see except needs to be closer to dark reddish pink
            Blue Team - Maybe could use slightly darker blue, other than that looks good that I can see

  - [ ] Nightshade
  - [ ] Full 64 bit compatibility
     - Front Grill
            Red Team - Needs more dark pink
            Blue Team - Needs to be very slightly a darker blue
  - Spikes
            Red Team - Needs more dark pink and missing a subtle glow (when I say glow I mean corona like so maybe not possible)
            Blue Team - Good, all it's missing now is that blue green Drakk like color pulse glow
  - Top Grill
            Red Team - Needs more dark pink (looks hot pink in my UT3 game but I'm on PS3 so TV color may differ from PC) and at the moment with the colors off I can't tell but looks like the pulse effect isn't quite right yet
  - Circle Lights
           Red Team - Hot pink in UT3 for me and more of an orange in UT2004 right now
           Blue Team - Needs to be darker blue, in UT3 the circle part most of the time is barely noticeable
  - Turret Lights
            Red Team - Needs a little more pink
  - Side Plating team color is missing
            Red Team - Where the spikes are above the round circle should be dark purple red plates but it's just black in UT2004 for me
            Blue Team - Where the spikes are above the round circle should be dark blue green plates but again it's currently just black

**Comment by Unre-Alex** [comment]
Whatever format works for you, since you will be the one to check out boxes.

After testing textures with HellDragon, here are the issues with 64bit:

Specular can not use a Combiner, TexEnvMap works fine.

Selfillumination can not use a Combiner or TexPanner (rip moving lights for the Necris vehicles), only a simple texture can be applied.

Other notes:

Diffuse does support multiple Combiners.

Any type of LODset for the lights works fine.

**Comment by GreatEmerald** [comment]
Thanks, it's useful to track this in a separate issue like this.

I can check off boxes as well. You should have a box before each "red team" and "blue team", and each "missing", i.e. the opposite of what you have right now.

Yes, that's about what I remember about it too. It's much more limiting than it could be, but it's better than having just the diffuse on 64-bit.

**Comment by HellDragon-HK** [comment]
Is that formatting any better GreatEmerald?

**Comment by GreatEmerald** [comment]
Yeap, that looks fine.

@Unre-Alex If you can make all textures compatible with 64-bit (and potentially solve the other issues in the OP), feel free to send a pull request with your texture package here.

**Comment by HellDragon-HK** [comment]
@Unre-Alex 
I'd like to apologize for angering you about needing to tweak stuff, I installed my PC UT3 and your light details are fairly accurate after all and my TV apparently is way off color, I'm putting together a screenshot pack to compare 32, 64, & UT3 right now, all I need to do is sort the UT3 screens and upload it

**Comment by Unre-Alex** [comment]
@GreatEmerald Yes, I will take care of all the issues related to the textures.

@HellDragon-HK No worries. So, do you want me to remove the lights from Paladin's back?

**Comment by HellDragon-HK** [comment]
I think that should be Emerald's call, it's inconsistent but it doesn't look bad really uh except for always glowing tail lights that is....yeah your say GreatEmerald

**Comment by HellDragon-HK** [comment]
Ok got the comparison screenshots put together finally so here you go, let me know if you need a different angle or a different map environment and I'll see what I can do.

https://drive.google.com/open?id=1nREoeiM1fNb_vhwQ1UnAcKB0rl4HSFbK

**Comment by Unre-Alex** [comment]
Thanks, I won't have to go back and forth with ut3 to check how they look. I didn't notice before that the lights per map have either a different color (like orange in ColdHarbor) or are darker/dull (like in Downtown), and they differ depending on what light environment they currently are on (which doesn't work in ut2k4) or even your screen's angle. The red lights look pretty the same in Islander as mine in UT2k4, the blue lights are brighter inside and darker outside tho. 

There's just no one way of how they look, so I will do what I think look best considering 64bit limits. Also, I will remove all the break lights, since they use coronas.

**Comment by HellDragon-HK** [comment]
If you need I can try to get multiple maps so you can try to find the common color....would turning off dynamic lighting for getting the screens help any you think?

**Comment by Unre-Alex** [comment]
I will focus on the bright maps like Suspense and Islander. In UT2k4 bright lights are better seen at distance. It's up to you what you will do.

**Comment by HellDragon-HK** [comment]
That's the opposite of what I was thinking, I thought dark maps would be easier because it wouldn't have the sunlight interference.

**Comment by Unre-Alex** [comment]
@100GPing100 Why did you set textures' NomalLod to 1 or 2 instead of the default 0? What does this do? I couldn't find anything about it on the unrealed wiki.

**Comment by 100GPing100** [comment]
@Unre-Alex With a `NormalLOD` of 1 or 2, at normal detail settings the miplevels drop to improve performance. This is important for lower end systems since these textures are at a high resolution.

**Comment by 100GPing100** [comment]
@Unre-Alex @HellDragon-HK If the difference in colours _between_ maps you guys are talking about is relative to UT3, this is because each map has a different post processing effect applied to the whole level (talking about stock maps here); it's this difference in post processing effects that is causing the difference in colours from map to map.

**Comment by GreatEmerald** [comment]
Yes, and the solution is to disable postprocessing (=bloom and depth of field). That said, the result tends to be highly desaturated, and I don't think that fits that well in UT2004...

And no, no need to remove lights from Paladin. My view is that if something is better in UT2004, it should be kept that way, and that if there are bugs in UT3 they should not be ported.

**Comment by HellDragon-HK** [comment]
Ok so glowing lights on the Paladin's rear side then....what about not having them glow and adding brakelights in the code for that area?  UT2004 doesn't have brakelights either that I can tell so it's not a bug but an addition....if that's going to be the case the Levi turret has a gray circle on it's main turret that looks like it was meant to have a light like the Hellbender can we add a light to it as the Levi doesn't have much

**Comment by GreatEmerald** [comment]
A brakelight makes more sense, yes. But anyway, that's off-topic to the texture discussion.

**Comment by HellDragon-HK** [comment]
Don't take this as a snarky comment but how is it off-topic when we are talking about needing textures added to add a brake light where there aren't any textures or lights originally?

If you're allowing creative freedom improvements over original UT3 then what about the making a light texture detail for the Levi driver turret like the Hellbender driver turret has, in UT3 the Levi doesn't have anything except a blank gray circle.    If yes then it's up to Alex if he wants to of course and it's just a detail I'd like to see.

**Comment by GreatEmerald** [comment]
Because we're not talking about textures, but rather coronas. Those don't need textures, they already got them.

Would be interesting to see how that looks like.

**Comment by HellDragon-HK** [comment]
None of the Paladins have a texture though :S, Alex added it where it didn't exist so that's a texture unless the texture is also in the code you mean....kind of confused on that now

**Comment by GreatEmerald** [comment]
Lights ought to be implemented as UT2004 Headlight/Brakelight actors, not via changing vehicle textures. That's the whole point.

**Comment by HellDragon-HK** [comment]
So then you want the texture removed and have a corona without the texture under it so it's just there? Really confused now GreatEmerald maybe I better just let Alex take over and continue this convo yeah?

**Comment by GreatEmerald** [comment]
The texture doesn't matter because the corona is right on top of it anyway.

**Comment by HellDragon-HK** [comment]
The detail matters when the vehicle is sitting empty and it has no coronas active though right?

**Comment by GreatEmerald** [comment]
I suppose. In that sense the lights should be off.

**Comment by HellDragon-HK** [comment]
You mean removed or a texture that looks like they are off?

**Comment by GreatEmerald** [comment]
The latter.

**Comment by Unre-Alex** [comment]
The Paladin's brake light is on its light texture. I will remove it, we don't wanna deal with people commenting on why it's there.

It does look like there was supposed to be a light on the Leviathan's driver turret, they either forgot to add it to the light texture or decided it's not needed for some reason. I could cut out the light from Hellbender's driver turret and past it there.

**Comment by HellDragon-HK** [comment]
Didn't GreatEmerald just say 'let's give the Paladin new brake lights & textures to improve it'?

Yeah that sounds good on the turret and one of the reasons they may have left the light off was to help PS3 performance as there are quite a few minor details like that missing on PS3 and maybe they thought the Levi is so rare who would notice (probably also what happened to the Scorps hatch animation because I don't think it even happens on PS3) although it could be also just have been forgotten like you said.

**Comment by HellDragon-HK** [comment]
Would custom headlight projectors be out of the question for ALL UT3 vehicles? They feel so much more alive with the projectors and it's something UT2004 does better over UT3 in my opinion.

With that said can all vehicles have projectors maybe excluding the Raptor and Cicada or making their front light texture area be the headlight?

**Comment by GreatEmerald** [comment]
Yes, I'm all for improved and consistent projectors, that's something UT2004 does better than UT3. Mind you, that should also be its own thread, this is about improved textures.

**Comment by HellDragon-HK** [comment]
Projectors are a texture though aren't they? :S

**Comment by GreatEmerald** [comment]
Projectors are actors that project a texture onto another texture. They are not textures themselves. The dicussion would be valid here if you wanted to introduce a new square texture for the projector (but I don't think you do, because you can reuse the Hellbender projector).

**Comment by HellDragon-HK** [comment]
I do mean texture, the Hellbender has 4 squares and also not oriented right for what we need, there isn't a single round circle texture to work for the Manta either....I don't know much about how it works I only know what I can see from testing in-game

**Comment by GreatEmerald** [comment]
OK then. Projectors are usually black-and-white textures, so making them shouldn't be hard.

**Comment by HellDragon-HK** [comment]
Alex and I were just discussing that in an e-mail before you came online, he is waiting for you to help with making team colored projectors I believe

**Comment by Unre-Alex** [comment]
Sure, if it's possible to add team colored projectors then I could try to make them by using headlights from UT3 textures. But this is the last thing I would like to do, because I don't care much about projectors.

**Comment by HellDragon-HK** [comment]
I know you don't want corona related things here but I don't know where else to put it so...
It looks like Coronas are already supposed to be team colored but it doesn't work?
class ONSHeadlightCorona extends Light
	native;

simulated function ChangeTeamTint(byte T)
{
	if(T == 0)
	{
		LightHue = 255;
		LightSaturation=240;
	}
	else
	{
		LightHue = 128;
		LightSaturation=175;
	}
}

If that was meant to work and doesn't would using a grey light with team textures work? Corona, Corona_0, Corona_1 maybe, it would also have a normal non team color in VDM or other non-team vehicle matches which would be nice.

**Comment by Unre-Alex** [comment]
You're right, they do work it's just they're set to be subtle. The LightHue and
LightSaturation values match in-game. If we could add this

simulated function ChangeTeamTint(byte T)
{
if(T == 0)
{
LightHue = 0;
LightSaturation=0;
}
else
{
LightHue = 175;
LightSaturation=0;
}
}

that would make a strong red and blue coronas.

**Comment by GreatEmerald** [comment]
Hm, yes, that's how it should work, but... If it simply does a `HeadlightCorona[i] = spawn( class'ONSHeadlightCorona'...`, then we can't directly influence its code. Several approaches could be used in this case:

- Replace this bit of code with one that spawns our custom `UT3HeadlightCorona` with the correct values;
- Use `CheckReplacement()` in a mutator to replace all `ONSHeadlightCorona`s with `UT3HeadlightCorona`s;
- Add new code to each vehicle to set the hue and saturation manually after each team switch (`TeamChanged`).

One big difference between these approaches would be network compatibility. The default headlight has `RemoteRole=ROLE_None`, which means it is completely client-side. I'm not sure if destroying it would be replicated or not... If not, you'd get two headlights on top of one another when using approach 2. Approach 3 would work as long as the code is run by the clients as well (it should be). Approach 1 would always work, but it requires copy-pasting a lot of code to each vehicle.

I'd probably go for option 3, and if it doesn't work, for option 1.

But yes, anyway, this is off-topic, open a new pull request or issue to discuss coronas.

**Comment by HellDragon-HK** [comment]
@GreatEmerald I already attempted something like that with the Manta I think, I posted it on the Manta issue because I was working with the Manta...guess it should have been here (wasn't thinking sorry)

**Comment by GreatEmerald** [comment]
Here's a more descriptive name for this issue to keep comments on topic. This should only concern @Unre-Alex's material replacements and nothing else (no coronas!).

**Comment by HellDragon-HK** [comment]
I think the Manta's headlight area texture needs to be changed to work better with the new team coronas, the white headlight with bright red or blue doesn't really make sense to me, anyone else agree?
